Avoid foreground services on API 34+

0nko opened the pull request:
This PR implements [a new user-initiated data-transfer](https://developer.android.com/about/versions/14/changes/user-initiated-data-transfers) `JobService` that avoids using foreground services. This is a [new policy requirement](https://support.google.com/googleplay/android-developer/answer/13315670) for devices running API 34+ from August 31, 2023 (p8Qyks-5tZ). Although this requirement applies only after we update the `targetSdk` to 34, I already started looking into it and implemented the change, so we might as well use it.

**To test:**
1. Create a new emulator running Android 14 (API 34)
2. Run the app and log in
3. Enable the notifications permission
4. Go to Products -> select a product
5. Tap on the new image button
6. Take a picture with a camera and confirm
7. Notice the image upload is initiated and a notification is shown
8. After the upload finishes, notice the notification is removed

0nko committed:
Update compileSdk to 34

0nko committed:
Add a new permission and JobService to the manifest

0nko committed:
Add new version check utility functions

0nko committed:
Add a new JobService for image uploads

0nko committed:
Add a new logic flow that uses the new JobService for Android 14

0nko commented:
@woocommerce/platform-9 👋 I've updated the `compileSdk` to 34 to use some new APIs but there's a problem with the SDK installation on the CI machine:

```
[2023-06-30T10:03:40Z] > Failed to install the following SDK components:
[2023-06-30T10:03:40Z]       platforms;android-34 Android SDK Platform 34
```
I'm not sure where to do this so I'd appreciate any help! 🙏

JorgeMucientes commented:
As stated in the previous comment from Ondrej, we can't merge this PR until we update the Buildkite CI environment to support Android SDK 34. Android core and platform team are aware of this and working on it :) 
More info at: p1689084678620839-slack-CGPNUU63E

JorgeMucientes committed:
Fix lint failing: make explicit with annotation the API level check

Lint should automatically detect that isAtLeastU() function is checking for the correct API level, but it seems there's a bug in the Lint latest Android version, so this workaround is needed for now to make Lint pass

JorgeMucientes commented:
Great job Ondrej! Everything works as described and the code looks good, so this is ready to be shipped.
